lib/hsa/mcwamp_hsa.cpp

```diff
@@ -4000,11 +4000,46 @@ hsa_status_t HSACopy::hcc_memory_async_copy(const Kalmar::HSADevice *copyDeviceA
         throw Kalmar::runtime_exception("copy agent must be GPU hcc_memory_async_copy", -1);
     }
 
-    // Pass same copy agent for source and device so we have control over the copy engine.
-    // The runtime will always pick the srcAgent if it is a GPU (which we checked above), and
-    // if at least on of the pointers is in device memory. (caller should ensure this is true).
+    hsa_agent_t hostAgent = const_cast<Kalmar::HSADevice *> (copyDeviceArg)->getHostAgent();
 
-    status = hsa_amd_memory_async_copy(dst, copyAgent, src, copyAgent, sizeBytes, depSignalCnt, depSignals, completion_signal);
+
+    hsa_agent_t srcAgent, dstAgent;
+    switch (copyKind) {
+        case Kalmar::hcMemcpyHostToHost: 
+            srcAgent=hostAgent; dstAgent=hostAgent;
+            //throw Kalmar::runtime_exception("HCC should not use hsa_memory_async_copy for host-to-host copy");
+            break;
+        case Kalmar::hcMemcpyHostToDevice: 
+            srcAgent=hostAgent; dstAgent=copyAgent;
+            break;
+        case Kalmar::hcMemcpyDeviceToHost: 
+            srcAgent=copyAgent; dstAgent=hostAgent;
+            break;
+        case Kalmar::hcMemcpyDeviceToDevice: 
+            srcAgent=copyAgent; dstAgent=copyAgent;
+            break;
+        default:
+            throw Kalmar::runtime_exception("bad copyKind in hcc_memory_async_copy", copyKind);
+    };
+
+
+    /* ROCR logic to select the copy agent:
+     *  
+     *  Decide which copy agent to use : 
+     *   
+     *   1. Pick source agent as long as the src agent is a GPU (regardless of the dst agent).
+     *   2. Pick destination agent only if the dst agent is a GPU. 
```

Comment:
just nickpicking here... should it be read as  "Pick destination agent **if only** the dst agent is a GPU"?